core/src/main/scala/io/projectglow/sql/SqlExtensionProvider.scala

```diff
@@ -43,130 +49,108 @@ class GlowSQLExtensions extends (SparkSessionExtensions => Unit) {
+    args.zipWithIndex.flatMap {
+      case (_arg: JMap[String, Any], idx: Int) =>
+        val arg = _arg.asScala
+        // If the argument is optional and doesn't have a matching input, don't add a new
+        // expression to the list of children.
+        if (arg.get("is_optional").exists(_.asInstanceOf[Boolean]) && idx >= exprs.size) {
+          None
+          // If we have a var args argument, the child expressions from here on are part of
+          // the var args list.
+        } else if (arg.get("is_var_args").exists(_.asInstanceOf[Boolean])) {
+          Some(exprs.slice(idx, exprs.size))
+        } else if (idx >= exprs.size) {
+          throw parameterError(functionName, exprs.size)
+        } else if (idx == args.size - 1 && exprs.size != args.size) {
+          throw parameterError(functionName, exprs.size)
         } else {
-          LiftOverCoordinatesExpr(exprs(0), exprs(1), exprs(2), exprs(3))
+          Some(exprs(idx))
         }
-      }
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("linear_regression_gwas"),
-      exprs => LinearRegressionExpr(exprs(0), exprs(1), exprs(2))
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("logistic_regression_gwas"),
-      exprs => LogisticRegressionExpr(exprs(0), exprs(1), exprs(2), exprs(3))
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("sample_call_summary_stats"),
-      exprs => CallSummaryStats(exprs(0), exprs(1), exprs(2))
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("sample_dp_summary_stats"),
-      exprs => rewrite(SampleDpSummaryStatistics(exprs.head))
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("sample_gq_summary_stats"),
-      exprs => rewrite(SampleGqSummaryStatistics(exprs.head))
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("subset_struct"),
-      exprs => rewrite(SubsetStruct(exprs.head, exprs.tail))
-    )
+    }
+  }
 
-    functionRegistry.registerFunction(
-      FunctionIdentifier("vector_to_array"),
-      exprs => VectorToArray(exprs.head)
-    )
+  /**
+   * Register SQL functions based on a yaml function definition file.
+   */
+  def registerFunctions(
+      conf: SQLConf,
+      functionRegistry: FunctionRegistry,
+      resourcePath: String = FUNCTION_YAML_PATH): Unit = {
+
+    loadFunctionDefinitions(resourcePath).foreach { _function =>
+      val function = _function.asScala
+      val id = FunctionIdentifier(function("name").asInstanceOf[String])
+      val exprClass = function("expr_class").asInstanceOf[String]
+      val args = function("args").asInstanceOf[JList[JMap[String, Any]]].asScala
+      val info = new ExpressionInfo(
+        exprClass,
+        null,
+        function("name").asInstanceOf[String],
+        function("doc").asInstanceOf[String],
+        makeArgsDoc(args),
+        "",
+        "",
+        function("since").asInstanceOf[String]
```

Comment:
Not urgent, but can we verify all of these are non-empty?

Reply:
Added a test